# A basic Principle Component Analysis from scratch on a small dataset:

In [3]:
import numpy as np 
np.random.seed(0)
X = np.random.randn(100, 4)   # 100 samples, 4 features
print("Dataset: \n",X)

Dataset: 
 [[ 1.76405235  0.40015721  0.97873798  2.2408932 ]
 [ 1.86755799 -0.97727788  0.95008842 -0.15135721]
 [-0.10321885  0.4105985   0.14404357  1.45427351]
 [ 0.76103773  0.12167502  0.44386323  0.33367433]
 [ 1.49407907 -0.20515826  0.3130677  -0.85409574]
 [-2.55298982  0.6536186   0.8644362  -0.74216502]
 [ 2.26975462 -1.45436567  0.04575852 -0.18718385]
 [ 1.53277921  1.46935877  0.15494743  0.37816252]
 [-0.88778575 -1.98079647 -0.34791215  0.15634897]
 [ 1.23029068  1.20237985 -0.38732682 -0.30230275]
 [-1.04855297 -1.42001794 -1.70627019  1.9507754 ]
 [-0.50965218 -0.4380743  -1.25279536  0.77749036]
 [-1.61389785 -0.21274028 -0.89546656  0.3869025 ]
 [-0.51080514 -1.18063218 -0.02818223  0.42833187]
 [ 0.06651722  0.3024719  -0.63432209 -0.36274117]
 [-0.67246045 -0.35955316 -0.81314628 -1.7262826 ]
 [ 0.17742614 -0.40178094 -1.63019835  0.46278226]
 [-0.90729836  0.0519454   0.72909056  0.12898291]
 [ 1.13940068 -1.23482582  0.40234164 -0.68481009]
 [-0.87079715 -0.578

In [8]:
# Compute the data by subtracting the column means.
col_mean=X.mean(axis=0)
print("Column means: ",col_mean)
X_centered=X-col_mean
print("Centered data: \n",X_centered)

Column means:  [ 0.02368218 -0.11178956 -0.1677182   0.1416176 ]
Centered data: 
 [[ 1.74037016  0.51194677  1.14645618  2.0992756 ]
 [ 1.84387581 -0.86548832  1.11780662 -0.29297481]
 [-0.12690103  0.52238806  0.31176177  1.31265591]
 [ 0.73735554  0.23346458  0.61158143  0.19205673]
 [ 1.47039689 -0.0933687   0.4807859  -0.99571334]
 [-2.576672    0.76540816  1.0321544  -0.88378262]
 [ 2.24607244 -1.34257611  0.21347672 -0.32880145]
 [ 1.50909703  1.58114833  0.32266562  0.23654492]
 [-0.91146793 -1.86900691 -0.18019395  0.01473137]
 [ 1.2066085   1.31416941 -0.21960862 -0.44392035]
 [-1.07223515 -1.30822837 -1.53855199  1.8091578 ]
 [-0.53333436 -0.32628474 -1.08507716  0.63587276]
 [-1.63758003 -0.10095072 -0.72774836  0.2452849 ]
 [-0.53448732 -1.06884262  0.13953597  0.28671427]
 [ 0.04283504  0.41426146 -0.46660389 -0.50435876]
 [-0.69614263 -0.2477636  -0.64542808 -1.8679002 ]
 [ 0.15374396 -0.28999137 -1.46248015  0.32116466]
 [-0.93098055  0.16373496  0.89680876 -0.01263469]


In [10]:
# Compute the covariance matrix: C = (X_centeredT @ X_centered)/(n-1).
cross=np.transpose(X_centered) @ X_centered
C=cross/(len(X_centered)-1)
print("The coveriance matrix: \n",C)


The coveriance matrix: 
 [[ 1.18362543 -0.07935745  0.13617526  0.07584143]
 [-0.07935745  0.95515849  0.17818152  0.05314995]
 [ 0.13617526  0.17818152  0.80104545  0.01782341]
 [ 0.07584143  0.05314995  0.01782341  0.94670617]]


In [11]:
# Get eigenvalues and eigen vectors using np.linalg.eigh()(symmetric matrix)
eigenvalues,eigenvectors=np.linalg.eigh(C)
print("Eigen Values: ",eigenvalues)
print("Eigen Vectors: \n",eigenvectors)

Eigen Values:  [0.63452451 0.91777943 1.08670055 1.24753104]
Eigen Vectors: 
 [[ 0.28821556  0.15704411  0.17368424 -0.92849487]
 [ 0.52924312  0.17423038 -0.82949171  0.03858752]
 [-0.78968318  0.3083845  -0.45197924 -0.27751438]
 [-0.11503905 -0.92188932 -0.27837394 -0.24370893]]


In [18]:
# Sort eigen values in descending order and reorder eigenvectors accordingly.

# Sorted index
idx=np.argsort(eigenvalues)[::-1]
# Reorder both eigenvalues and eigenvectors
eigvals_sorted = eigenvalues[idx]
eigvecs_sorted = eigenvectors[:, idx]
print("Sorted Eigen Values: \n",eigvals_sorted)
print("Sorted Eigen Vectors:\n",eigvecs_sorted)

Sorted Eigen Values: 
 [1.24753104 1.08670055 0.91777943 0.63452451]
Sorted Eigen Vectors:
 [[-0.92849487  0.17368424  0.15704411  0.28821556]
 [ 0.03858752 -0.82949171  0.17423038  0.52924312]
 [-0.27751438 -0.45197924  0.3083845  -0.78968318]
 [-0.24370893 -0.27837394 -0.92188932 -0.11503905]]


In [ ]:
# Compute the explained variance ratio for each componenet: 
eigval_sum=eigvals_sorted.sum()
variance_ratio=eigvals_sorted/eigval_sum
print("Explained variance ration for each component: \n",variance_ratio)

Explained variance ration for each component: 
 [0.32098794 0.27960649 0.23614333 0.16326224]


In [22]:
# Project the data onto the top 2 components - report the shape of the result.
projection=X_centered @ eigenvectors[:,:2]
print("The top 2 components: \n",projection)
print("Shape: ",projection.shape)

The top 2 components: 
 [[-0.37428978 -1.21923888]
 [-0.77562959  0.75358005]
 [-0.15730428 -1.04289416]
 [-0.16897274  0.16802115]
 [ 0.10925389  1.28085391]
 [-1.05095541  0.86175638]
 [-0.19395012  0.48776635]
 [ 0.98974159  0.39391571]
 [-1.11125683 -0.53792817]
 [ 1.26776787  0.75998052]
 [ 0.00543912 -2.53863018]
 [ 0.45731778 -1.06143102]
 [ 0.02106979 -0.72531282]
 [-0.86289771 -0.49145104]
 [ 0.65808095  0.39997347]
 [ 0.39279885  1.37046419]
 [ 1.00878496 -0.77346521]
 [-0.88840879  0.1705323 ]
 [-0.62788699  0.91722246]
 [-0.3815773  -0.18742752]
 [-0.11387028  1.73185659]
 [ 0.45834531  1.2914663 ]
 [ 0.36339249 -1.03769501]
 [ 0.15039312 -0.1405804 ]
 [ 0.73790671  0.17938532]
 [ 0.65740888 -1.02652178]
 [ 1.03933576  0.9139581 ]
 [-0.30505076  0.49862615]
 [ 0.81817962 -0.42684771]
 [-0.46385761 -0.34101162]
 [-0.92527286 -1.06520714]
 [-0.21229682 -1.77612433]
 [ 0.89155284 -0.35995066]
 [ 0.18297925 -0.72200039]
 [-0.19484773  1.38209772]
 [-0.46574568 -0.49363157]
 [ 1

# Monthly returns for 4 assets over 24 months.

In [24]:
np.random.seed(42)
returns = np.random.randn(24, 4) * 0.05 + 0.01
# shape (24 months, 4 assets)
print("Raw\n",returns)

Raw
 [[ 0.03483571  0.00308678  0.04238443  0.08615149]
 [-0.00170767 -0.00170685  0.08896064  0.04837174]
 [-0.01347372  0.037128   -0.01317088 -0.01328649]
 [ 0.02209811 -0.08566401 -0.07624589 -0.01811438]
 [-0.04064156  0.02571237 -0.0354012  -0.06061519]
 [ 0.08328244 -0.00128882  0.01337641 -0.06123741]
 [-0.01721914  0.01554613 -0.04754968  0.0287849 ]
 [-0.02003193 -0.00458469 -0.02008533  0.10261391]
 [ 0.00932514 -0.04288555  0.05112725 -0.05104218]
 [ 0.02044318 -0.08798351 -0.0564093   0.01984306]
 [ 0.04692333  0.01856841  0.00421759 -0.00505518]
 [-0.0639261  -0.02599221 -0.01303194  0.06285611]
 [ 0.02718091 -0.07815201  0.0262042  -0.00925411]
 [-0.0238461   0.04058381  0.06154998  0.05656401]
 [-0.03196088 -0.00546062  0.02656317  0.05877726]
 [-0.01395871  0.00071705 -0.04531675 -0.04981033]
 [ 0.05062629  0.077812    0.00639949  0.06017664]
 [ 0.0280818  -0.02225599  0.02806978  0.08690183]
 [ 0.0082087   0.08823218 -0.12098726  0.05109513]
 [ 0.01435235 -0.00495037 

In [25]:
# Compute the covariance matrix of asset returns(use np.cov(returns.T))
returnsT=np.transpose(returns)
cov_returns=np.cov(returnsT)
print("Covariance matrix of asset returns: \n",cov_returns)


Covariance matrix of asset returns: 
 [[ 1.12416481e-03 -9.61277983e-05  5.94262121e-05 -2.49202220e-04]
 [-9.61277983e-05  1.94394193e-03 -3.66154140e-05  4.61276110e-04]
 [ 5.94262121e-05 -3.66154140e-05  2.56022676e-03  2.52738440e-04]
 [-2.49202220e-04  4.61276110e-04  2.52738440e-04  3.10296299e-03]]


In [28]:
# Compute eigenvalues of the covariance matrix - these represent reisk componenets.
eig_vals=np.linalg.eigvalsh(cov_returns)
print("Eigen values of the covariance matrix are: \n",eig_vals)

Eigen values of the covariance matrix are: 
 [0.0010862  0.00176457 0.00252331 0.00335721]


In [30]:
# Find what percentage of total risk(sum of eigenvalues) is explained by the largest eigenvalue. 
percentage_risk=(eig_vals.max()/eig_vals.sum())*100
print(f"The percentage of total risk explained by the largest eigenvalue({np.round(eig_vals.max(),5)}) is {np.round(percentage_risk,5)}%")

The percentage of total risk explained by the largest eigenvalue(0.00336) is 38.45035%


Create an equal-weight portfolio vector w=[0.25,0.25,0.25,0.25]
Compute portfolio variance and report portfolio std

In [36]:
w=np.array([0.25,0.25,0.25,0.25])
wT=np.transpose(w)

portfolio_var=wT @ cov_returns @ w
print("Portfolio variance: ",np.round(portfolio_var,6))
print("Portfolio standard deviation: ",np.round(np.sqrt(portfolio_var),6))



Portfolio variance:  0.000595
Portfolio standard deviation:  0.024385


# A simple Linear Regression solver from scratch using the Normal Equation. Dataset: house sizes and prices.

In [57]:
# House sizes (sq ft) and prices ($1000s)
sizes  = np.array([1400, 1600, 1700, 1875, 1100,
                   1550, 2350, 2450, 1425, 1700], dtype=float)
prices = np.array([245, 312, 279, 308, 199,
                   219, 405, 324, 319, 255], dtype=float)
# np.set_printoptions(suppress=True)   # suppress scientific notation
# np.set_printoptions(precision=6)

In [ ]:
# Build the design matrix 'X' by stacking a column of ones(bias term) with the sizes column.
ones=np.ones(10)
X = np.column_stack([ones, sizes])
print("Matrix X:\n", X)

Matrix X:
 [[   1. 1400.]
 [   1. 1600.]
 [   1. 1700.]
 [   1. 1875.]
 [   1. 1100.]
 [   1. 1550.]
 [   1. 2350.]
 [   1. 2450.]
 [   1. 1425.]
 [   1. 1700.]]


In [60]:
# Compute weights using the Normal Equaltion: w = inv(XT X) @ XT y. where y is prices
XT=np.transpose(X)
dot_xtx=np.dot(XT,X)
dot_xty=np.dot(XT,prices)

w=np.linalg.inv(dot_xtx) @ dot_xty
print("Weights: \n",w)


Weights: 
 [98.24833   0.109768]


* Intercept (bias term) = 98.25  
→ This is the baseline price (in $1000s) when the house size is 0. Of course, a house can’t be 0 sq ft, but mathematically it’s the starting point of the line.

* Slope (coefficient for size) = 0.1098  
→ For each additional square foot, the price increases by about 0.11 thousand dollars, i.e. $110 per sq ft.

In [61]:
# Verify det(XT X)!=0 before inverting.
det_xtx=np.linalg.det(dot_xtx)
print("Determinant of dot product of XT & X: ",det_xtx)

Determinant of dot product of XT & X:  15715000.000000004


Since the determinant doesn't equals to zero. So the system has unique solution.

In [66]:
# Use the weights to predict prices for the training data: y_pred=X @ w.
y_pred=(X @ w)
y_pred_view=y_pred.reshape(-1,1)*1000
print("Predicted prices for the training data(in $): \n",y_pred_view)

Predicted prices for the training data(in $): 
 [[251923.162584]
 [273876.71015 ]
 [284853.483933]
 [304062.838053]
 [218992.841234]
 [268388.323258]
 [356202.513522]
 [367179.287305]
 [254667.356029]
 [284853.483933]]


In [69]:
# Compute Mean Squared Error: MSE= mean((y-y_pred)^2)
MSE=np.mean(np.square(prices-y_pred))
print("Mean Squared Error: ",MSE)

RMSE=np.sqrt(MSE)
print("Root Mean Square Error: ",RMSE)

Mean Squared Error:  1366.5565224307989
Root Mean Square Error:  36.96696528565469


* Mean Squared Error (MSE) = 1366.56  
→ This is the average of the squared prediction errors. It’s in units of ($1000)^2, so less intuitive on its own.
* Root Mean Squared Error (RMSE) ≈ 37  
→ On average, the model’s predictions are off by about $37,000 compared to actual house prices.

In [76]:
# Find which house has the largest prediction error using argmax() on the absolute error.
abs_error=prices-y_pred;
max_error_houseIdx=np.argmax(abs_error)
print(f"House no. {max_error_houseIdx+1} has the largest prediction error of roughy ${np.round(np.max(abs_error)*1000,0)}\n where actual price was ${prices[max_error_houseIdx]*1000} but system predicted ${np.round(y_pred[max_error_houseIdx]*1000,0)}")

House no. 9 has the largest prediction error of roughy $64333.0
 where actual price was $319000.0 but system predicted $254667.0


In [ ]:
# Verify the Normal Equation solution matches np.linalg.solve(XT @ X,XT @ prices).
w2=np.linalg.solve(XT @ X,XT @ prices)
print("New Normal Equaltion solution: \n",w2)
print("Do (inv(XT X) @ XT prices) and np.linalg.solve(XT @ X,XT @ prices) matche?: ",np.allclose(w,w2))

New Normal Equaltion solution: 
 [98.24833   0.109768]
Do (inv(XT X) @ XT prices) and np.linalg.solve(XT @ X,XT @ prices) matches?:  True


* Both methods give the same regression weights — intercept ≈ 98.25 and slope ≈ 0.1098 — confirming that the Normal Equation solution is consistent whether you use the matrix inverse or np.linalg.solve.